core/src/main/java/org/bitcoinj/core/Block.java

```diff
@@ -190,15 +193,36 @@ public Block(NetworkParameters params, byte[] payloadBytes, int offset, @Nullabl
      */
     public Block(NetworkParameters params, long version, Sha256Hash prevBlockHash, Sha256Hash merkleRoot, long time,
                  long difficultyTarget, long nonce, List<Transaction> transactions) {
+        this(params, version, prevBlockHash, merkleRoot, time, difficultyTarget, nonce, transactions, null);
+    }
+
+    // private for now, but maybe we'll make public later - full non-deserializing constructor
+    // this one can copy a hash to avoid recomputing
+    private Block(NetworkParameters params, long version, Sha256Hash prevBlockHash, Sha256Hash merkleRoot, long time,
+                 long difficultyTarget, long nonce, List<Transaction> transactions, Sha256Hash hash) {
         super(params);
         this.version = version;
         this.prevBlockHash = prevBlockHash;
         this.merkleRoot = merkleRoot;
         this.time = time;
         this.difficultyTarget = difficultyTarget;
         this.nonce = nonce;
-        this.transactions = new LinkedList<>();
-        this.transactions.addAll(transactions);
+        this.hash = hash;
+        if (transactions != null) {
```

Comment:
Good question. Generally its a best practice to use empty lists instead of null lists.

My (probably flawed) thinking was that `transactions == null` for header-only, and that in a real block transactions would be present. Of course if `Block` were truly immutable there would never be a valid block without at least one transaction, right?

Reply:
Yes, the comment for `transactions`  says `// If null, it means this object holds only the headers.` -- but maybe we should change that (probably later, though.)